### PR TITLE
Autopilot GKE documentation for Elastic Agent

### PR DESCRIPTION
**Pull request by gizas**
## What does this PR do?

This PR adds documentation for the installation of Elastic Agent in GKE Autopilot cluster, for both Managed and standalone scenarions

## Why is it important?

Adds the needed manifests for Kubernetes Integration under a specific Cloud Managed Scenario which is the GKE Autopilot one. 

## Checklist

- [x] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

- Relates https://github.com/elastic/integrations/issues/4620

### CODE REVIEW

File: docs/elastic-agent-gke-autopilot.md
Comment (ChrsMark) [inline docs/elastic-agent-gke-autopilot.md:3]:
```suggestion
The goal of this document is to assist users to Kubernetes Observability of Autopilot Google Kubernetes Engine Clusters (Autopilot GKE) with Elastic Agent. Monitoring, observing and alerting of a Google Kubernetes Cluster made easy with the following simple steps. Elastic Agent is installed as daemonset inside Google Kubernetes Cluster and will send metrics/logs to a configured Elastic Cloud instance.
```

File: docs/elastic-agent-gke-autopilot.md
Comment (ChrsMark) [inline docs/elastic-agent-gke-autopilot.md:5]:
```suggestion
> Elastic Agent can be configured with multiple output setups, as part of this guide we will use Elastic Cloud
```

File: docs/elastic-agent-gke-autopilot.md
Comment (ChrsMark) [inline docs/elastic-agent-gke-autopilot.md:161]: Is this `_redused` suffix actually meaningful at this point?

File: docs/elastic-agent-gke-autopilot.md
Comment (ChrsMark) [inline docs/elastic-agent-gke-autopilot.md:185]: 👍🏼

File: docs/manifests/elastic-agent-managed-gke_autopilot_redused.yaml
Comment (ChrsMark) [inline docs/manifests/elastic-agent-managed-gke_autopilot_redused.yaml:56]: This is not needed right? If so we can remove it completely for clarity.

File: docs/manifests/elastic-agent-managed-gke_autopilot_redused.yaml
Comment (ChrsMark) [inline docs/manifests/elastic-agent-managed-gke_autopilot_redused.yaml:59]: Same for this one and the rest that are not useful at this point.

File: docs/manifests/elastic-agent-managed-gke_autopilot_redused.yaml
Comment (ChrsMark) [inline docs/manifests/elastic-agent-managed-gke_autopilot_redused.yaml:33]: Maybe remove those completely, same as the rest?

File: docs/elastic-agent-gke-autopilot.md
Comment (gizas) [reply docs/elastic-agent-gke-autopilot.md:161]: Hmm you are right, not so much. I will remove them

File: docs/manifests/elastic-agent-managed-gke_autopilot_redused.yaml
Comment (gizas) [reply docs/manifests/elastic-agent-managed-gke_autopilot_redused.yaml:56]: It is needed. Why? For the managed agent you need to update the token

File: docs/manifests/elastic-agent-managed-gke_autopilot_redused.yaml
Comment (gizas) [reply docs/manifests/elastic-agent-managed-gke_autopilot_redused.yaml:59]: Ah you mean the host and fleet username? I will remove them np

File: docs/manifests/elastic-agent-managed-gke_autopilot_redused.yaml
Comment (gizas) [reply docs/manifests/elastic-agent-managed-gke_autopilot_redused.yaml]: Done

**Review by ChrsMark — APPROVED** [review]
lgtm! @gizas can you add a more specific description in the PR's description so as we know why is this added and what is the plan forward to move it to the docs pages etc? In this, we will know all the details by looking into the history of the repo :)

File: docs/elastic-agent-gke-autopilot.md
Comment (kuisathaverat) [inline docs/elastic-agent-gke-autopilot.md:88]:
```suggestion
The Kubernetes Integration is an additional package to be installed in your Elastic Cloud and will enhance user experience for Kubernetes Observability by installing additional assets in our Elastic stack.
```

File: docs/elastic-agent-gke-autopilot.md
Comment (gizas) [reply docs/elastic-agent-gke-autopilot.md:152]: Yes like an example. 
Do you want to put something else more clear?

File: docs/elastic-agent-gke-autopilot.md
Comment (kuisathaverat) [inline docs/elastic-agent-gke-autopilot.md:183]: I think is better to recommend install a stable version from the Helm Chart, that allows to set the namespace to install.

https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-state-metrics

File: docs/elastic-agent-gke-autopilot.md
Comment (kuisathaverat) [reply docs/elastic-agent-gke-autopilot.md:152]: The thing is that it said, `The Elasticsearch host to communicate with` but it is a Kibana URL, which confuses me.

File: docs/scripts/autopilot-tests.sh
Comment (gizas) [reply docs/scripts/autopilot-tests.sh:5]: In. autopilot we are not able to install anything in kube-system. 
That is why we need to create a new namespace

File: docs/elastic-agent-gke-autopilot.md
Comment (gizas) [reply docs/elastic-agent-gke-autopilot.md]: done with latest commit

File: docs/elastic-agent-gke-autopilot.md
Comment (gizas) [reply docs/elastic-agent-gke-autopilot.md]: updated

File: docs/scripts/autopilot-tests.sh
Comment (gizas) [reply docs/scripts/autopilot-tests.sh:5]: Indeed that is why I have added them as variables. Do you want to provide them as arguments? 
I guess for now is ok

Comment (kuisathaverat) [inline docs/scripts/autopilot-tests.sh:5]:
```suggestion
NAMESPACE=${1:-"kube-system"}
```

File: docs/scripts/autopilot-tests.sh
Comment (kuisathaverat) [inline docs/scripts/autopilot-tests.sh:7]:
```suggestion
LEADERAGENT=${2-"elastic-agent-${LEADER}"}
```

File: docs/manifests/elastic-agent-standalone-gke-autopilot.yaml
Comment (kuisathaverat) [inline docs/manifests/elastic-agent-standalone-gke-autopilot.yaml:27]:
```suggestion
      - name: kubernetes-0
         id: C07CC023-84F5-4623-962C-5F82E9DF7841
```

File: docs/manifests/elastic-agent-standalone-gke-autopilot.yaml
Comment (kuisathaverat) [inline docs/manifests/elastic-agent-standalone-gke-autopilot.yaml:248]:
```suggestion
      - name: kubernetes-2
         id: C07CC023-84F5-4623-962C-5F82E9DF7888
```

File: docs/manifests/elastic-agent-standalone-gke-autopilot.yaml
Comment (kuisathaverat) [inline docs/manifests/elastic-agent-standalone-gke-autopilot.yaml:267]:
```suggestion
      - name: kubernetes-3
         id: C07CC023-84F5-4623-962C-5F82E9DF7877
```

File: docs/manifests/elastic-agent-standalone-gke-autopilot.yaml
Comment (kuisathaverat) [inline docs/manifests/elastic-agent-standalone-gke-autopilot.yaml:281]:
```suggestion
      - name: kubernetes-4
         id: C07CC023-84F5-4623-962C-5F82E9DF7866
```

File: docs/manifests/elastic-agent-standalone-gke-autopilot.yaml
Comment (kuisathaverat) [inline docs/manifests/elastic-agent-standalone-gke-autopilot.yaml:296]:
```suggestion
      - name: kubernetes-5
         id: C07CC023-84F5-4623-962C-5F82E9DF7855
```

File: docs/manifests/elastic-agent-standalone-gke-autopilot.yaml
Comment (kuisathaverat) [reply docs/manifests/elastic-agent-standalone-gke-autopilot.yaml:27]: this fix the issue with duplicates inputs IDs

**Comment by ChrsMark** [comment]
Hey folks, I was executing some manual tests and falling on some issues. I'm sharing here for now but maybe we will need a separate issue for those:


### Leader Agent is falling behind in collecting from kube-state-metrics 
In a very basic scenario I observed that ksm metrics are coming into Elastic really rarely. By looking into the logs of the leader Agent I notice:

```
{"log.level":"info","@timestamp":"2023-04-01T11:15:24.970Z","log.origin":{"file.name":"coordinator/coordinator.go","file.line":933},"message":"Unit state changed kubernetes/metrics-default-kubernetes/metrics-kube-state-metrics-f98f956c-bc0c-4c78-98cd-c36377aae5bd (STARTING->STOPPED): Suppressing FAILED state due to restart for '14011' exited with code '-1'","log":{"source":"elastic-agent"},"component":{"id":"kubernetes/metrics-default","state":"STOPPED"},"unit":{"id":"kubernetes/metrics-default-kubernetes/metrics-kube-state-metrics-f98f956c-bc0c-4c78-98cd-c36377aae5bd","type":"input","state":"STOPPED","old_state":"STARTING"},"ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-04-01T11:15:26.574Z","log.origin":{"file.name":"coordinator/coordinator.go","file.line":921},"message":"Spawned new unit kubernetes/metrics-default-kubernetes/metrics-kube-state-metrics-f98f956c-bc0c-4c78-98cd-c36377aae5bd: Starting: spawned pid '14035'","log":{"source":"elastic-agent"},"component":{"id":"kubernetes/metrics-default","state":"STARTING"},"unit":{"id":"kubernetes/metrics-default-kubernetes/metrics-kube-state-metrics-f98f956c-bc0c-4c78-98cd-c36377aae5bd","type":"input","state":"STARTING"},"ecs.version":"1.6.0"}
```

This makes me assume that Autopilot is providing the same resources' limit in the Daemonset Agents and as a result the leader is not functioning properly.

In order to fix this I remove the ksm data_streams from the policy and I create a new one called cluster-policy which I assing to a Deployment:
```yml
# For more information refer to https://www.elastic.co/guide/en/fleet/current/running-on-kubernetes-managed-by-fleet.html
apiVersion: apps/v1
kind: Deployment
metadata:
  name: elastic-agent-cluster
  namespace: elastic-agent
  labels:
    app: elastic-agent-cluster
spec:
  replicas: 1
  selector:
    matchLabels:
      app: elastic-agent-cluster
  template:
    metadata:
      labels:
        app: elastic-agent-cluster
    spec:
      # Tolerations are needed to run Elastic Agent on Kubernetes control-plane nodes.
      # Agents running on control-plane nodes collect metrics from the control plane components (scheduler, controller manager) of Kubernetes
      tolerations:
        - key: node-role.kubernetes.io/control-plane
          effect: NoSchedule
        - key: node-role.kubernetes.io/master
          effect: NoSchedule
      serviceAccountName: elastic-agent
      dnsPolicy: ClusterFirstWithHostNet
      containers:
        - name: elastic-agent-cluster
          image: docker.elastic.co/beats/elastic-agent:8.7.0
          env:
            # Set to 1 for enrollment into Fleet server. If not set, Elastic Agent is run in standalone mode
            - name: FLEET_ENROLL
              value: "1"
            # Set to true to communicate with Fleet with either insecure HTTP or unverified HTTPS
            - name: FLEET_INSECURE
              value: "true"
            # Fleet Server URL to enroll the Elastic Agent into
            # FLEET_URL can be found in Kibana, go to Management > Fleet > Settings
            - name: FLEET_URL
              value: "https://lakislalakis.fleet.europe-west1.gcp.cloud.es.io:443"
            # Elasticsearch API key used to enroll Elastic Agents in Fleet (https://www.elastic.co/guide/en/fleet/current/fleet-enrollment-tokens.html#fleet-enrollment-tokens)
            # If FLEET_ENROLLMENT_TOKEN is empty then KIBANA_HOST, KIBANA_FLEET_USERNAME, KIBANA_FLEET_PASSWORD are needed
            - name: FLEET_ENROLLMENT_TOKEN
              value: "plwplspwlpwls=="
            - name: NODE_NAME
              valueFrom:
                fieldRef:
                  fieldPath: spec.nodeName
            - name: POD_NAME
              valueFrom:
                fieldRef:
                  fieldPath: metadata.name
          securityContext:
            runAsUser: 0
          resources:
            limits:
              memory: 900Mi
              ephemeral-storage: "4Gi"
            requests:
              memory: 600Mi
              ephemeral-storage: "2Gi"
          volumeMounts:
            - name: varlog
              mountPath: /var/log
              readOnly: true
      volumes:
        - name: varlog
          hostPath:
            path: /var/log
```

The result is that we end up with one additional single Agent for cluster level metrics collection:

```console
NAME                                     READY   STATUS    RESTARTS   AGE
elastic-agent-4bs5f                      1/1     Running   0          20m
elastic-agent-cluster-768f998dbc-p8qv2   1/1     Running   0          2m14s
elastic-agent-scq59                      1/1     Running   0          20m
elastic-agent-w9zpn                      1/1     Running   0          10m
kube-state-metrics-565495c7d6-qll6n      1/1     Running   0          34m
```

With this we can bring the monitoring to a more stable phase and collect all the metrics without misses. Apparently, without proper feed of ksm metrics our out-of-the-box metrics do not work well since we have serious gaps in the timeseries graphs.


### Agent of Deamonset sometimes gets into Pending phase

Some times one Agent Pod can go into Pending state. If we check the Pod with `describe` we see the following:

```
Events:
  Type     Reason             Age    From                                   Message
  ----     ------             ----   ----                                   -------
  Warning  FailedScheduling   3m45s  gke.io/optimize-utilization-scheduler  0/3 nodes are available: 1 Insufficient cpu, 1 Insufficient memory. preemption: 0/3 nodes are available: 3 No preemption victims found for incoming pod.
  Normal   NotTriggerScaleUp  3m42s  cluster-autoscaler                     pod didn't trigger scale-up (it wouldn't fit if a new node is added): 2 node(s) didn't match Pod's node affinity/selector
``` 

In my case I was observing that the Daemonset Pod could not be scheduled on the node where KSM Pod was running.
To tackle this I created a priority policy:

```yml
apiVersion: scheduling.k8s.io/v1
kind: PriorityClass
metadata:
  name: elastic-agent
value: 1000000000
preemptionPolicy: PreemptLowerPriority
globalDefault: false
description: "Ensure that Elastic Agent Pods are always scheduled onto Nodes, by evicting other non-essential workloads."
``` 

and then assign this to the Deamonset Pods:

```yml
      ...
      serviceAccountName: elastic-agent
      dnsPolicy: ClusterFirstWithHostNet
      priorityClassName: elastic-agent
      containers:
        - name: elastic-agent
     ...
```

If we restart the Daemonset we ensure that all the Daemonset Pods will be scheduled first. However, it's still not clear to me what resources we need to define for the Daemonset so as to avoid having Nodes that wouldn't fit a lot of extra workload Pods.


@gizas @kuisathaverat have you ever  noticed such issues? 
I think the best approach here will be to create follow-up issue to particularly deal with those 2 issues.

**Comment by gizas** [comment]
> @gizas @kuisathaverat have you ever noticed such issues?

https://github.com/elastic/observability-test-environments/issues/4214#issuecomment-1486740736

So I am listing the following:
- Documentation/examples for deployment setup
- Documenting PriorityClass
- The resource calculation can be based on the parallel effort we do with K8s-scaling. So to calculate some resource numbers from there (based on specific number of pods running). Some numbers for 8.5 can also be found [here](https://docs.google.com/document/d/1B0zJmJTdtWyG4uo5g1dAT9PReEabH88bxIyrR9u4Qmo/edit)

**Comment by kuisathaverat** [comment]
@ChrsMark 
>This makes me assume that Autopilot is providing the same resources' limit in the Daemonset Agents and as a result the leader is not functioning properly.

Limits on Autopilot have a trick, you can not pass a request and a limit only one will be applied, see [Autopilot resource-limits](https://cloud.google.com/kubernetes-engine/docs/concepts/autopilot-resource-requests#resource-limits)

File: docs/scripts/autopilot-tests.sh
Comment (v1v) [inline docs/scripts/autopilot-tests.sh:83]: Will `exit $STATUS` be required to populate the error?

Comment (gizas) [reply docs/scripts/autopilot-tests.sh:83]: Well I used the $STATUS because can also be handy for your jenkins job to mark the status of your jenkins job as well.

We can decide what values you want me to return in any case and accordingly to set the jenkins status

**Comment by ChrsMark** [comment]
> @ChrsMark
> 
> > This makes me assume that Autopilot is providing the same resources' limit in the Daemonset Agents and as a result the leader is not functioning properly.
> 
> Limits on Autopilot have a trick, you can not pass a request and a limit only one will be applied, see [Autopilot resource-limits](https://cloud.google.com/kubernetes-engine/docs/concepts/autopilot-resource-requests#resource-limits)

Thanks @kuisathaverat ! I will run more tests to check how this work out.

@gizas I guess you can add the `PriorityClass`  in the documented manifests anyways, right?

Here is what I used:

```yml
apiVersion: scheduling.k8s.io/v1
kind: PriorityClass
metadata:
  name: elastic-agent
value: 1000000000
preemptionPolicy: PreemptLowerPriority
globalDefault: false
description: "Ensure that Elastic Agent Pods are always scheduled onto Nodes, by evicting other non-essential workloads."
```

and then referenced this from within Agent's manifest:

```yml
...
serviceAccountName: elastic-agent
dnsPolicy: ClusterFirstWithHostNet
priorityClassName: elastic-agent
...
```

File: docs/manifests/elastic-agent-managed-gke-autopilot.yaml
Comment (gizas) [reply docs/manifests/elastic-agent-managed-gke-autopilot.yaml]: done

**Comment by gizas** [comment]
@ChrsMark latest updates for PriorityClass and cpu commenting done. Please have a final review and we can close this.

Related to this https://github.com/elastic/elastic-agent/pull/1961#issuecomment-1493836302 comment: 
1. I will need to update our documentation for autopilot GKE cluster.
2. Relevant examples for deployments and agent resource will be tracked as part of the separate effort for kubernetes scaling project

**Comment by gizas** [comment]
@kuisathaverat if no objections I will merge this after final commits. Do you think I am missing anything?

**Pull request merged**
